File: CartaoFidelidade/Services/CnpjService.swift
//
//  CnpjService.swift
//  CartaoFidelidade
//
//  Validação de CNPJ via BrasilAPI
//  https://brasilapi.com.br/api/cnpj/v1/{cnpj}
//

import Foundation

struct CnpjValidationResult {
    let valid: Bool
    let razaoSocial: String?
    let error: String?
}

@MainActor
final class CnpjService {
    static let shared = CnpjService()
    private let baseURL = "https://brasilapi.com.br/api/cnpj/v1"

    private init() {}

    /// Valida CNPJ consultando a BrasilAPI
    func validate(cnpj: String) async -> CnpjValidationResult {
        let digits = cnpj.filter { $0.isNumber }
        guard digits.count == 14 else {
            return CnpjValidationResult(valid: false, razaoSocial: nil, error: "CNPJ deve ter 14 dígitos")
        }

        guard let url = URL(string: "\(baseURL)/\(digits)") else {
            return CnpjValidationResult(valid: false, razaoSocial: nil, error: "URL inválida")
        }

        do {
            let (_, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse else {
                return CnpjValidationResult(valid: false, razaoSocial: nil, error: "Resposta inválida")
            }
            if http.statusCode == 200 {
                return CnpjValidationResult(valid: true, razaoSocial: nil, error: nil)
            }
            return CnpjValidationResult(valid: false, razaoSocial: nil, error: "CNPJ não encontrado")
        } catch {
            print("❌ [CnpjService] Erro ao validar CNPJ: \(error.localizedDescription)")
            return CnpjValidationResult(valid: false, razaoSocial: nil, error: "Erro ao consultar CNPJ")
        }
    }
}

File: src/services/cnpjService.ts
/**
 * Serviço de validação de CNPJ via BrasilAPI
 * https://brasilapi.com.br/api/cnpj/v1/{cnpj}
 */

const BRASIL_API_CNPJ = "https://brasilapi.com.br/api/cnpj/v1";

export type CnpjValidationResult =
  | { valid: true; razaoSocial?: string }
  | { valid: false; error?: string };

/**
 * Valida CNPJ consultando a BrasilAPI.
 * Retorna { valid: true } se o CNPJ existe na Receita Federal.
 * Retorna { valid: false } se não encontrado ou erro na API.
 */
export async function validateCnpj(cnpj: string): Promise<CnpjValidationResult> {
  const digits = cnpj.replace(/\D/g, "");
  if (digits.length !== 14) {
    return { valid: false, error: "CNPJ deve ter 14 dígitos" };
  }

  try {
    const res = await fetch(`${BRASIL_API_CNPJ}/${digits}`);
    if (res.ok) {
      const data = (await res.json()) as { razao_social?: string };
      return { valid: true, razaoSocial: data.razao_social };
    }
    return { valid: false, error: "CNPJ não encontrado" };
  } catch (e) {
    console.warn("[cnpjService] Erro ao validar CNPJ:", e);
    return { valid: false, error: "Erro ao consultar CNPJ" };
  }
}
